rusb-async/src/pool.rs

```diff
@@ -0,0 +1,183 @@
+use std::collections::VecDeque;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use rusb::{ffi, DeviceHandle, UsbContext};
+
+use crate::{error::Error, error::Result, Transfer};
+
+use std::sync::atomic::{AtomicBool, Ordering};
+
+/// Represents a pool of asynchronous transfers, that can be polled to completion
+pub struct TransferPool<C: UsbContext> {
+    device: Arc<DeviceHandle<C>>,
+    pending: VecDeque<Transfer>,
+}
+
+impl<C: UsbContext> TransferPool<C> {
+    pub fn new(device: Arc<DeviceHandle<C>>) -> Result<Self> {
+        Ok(Self {
+            device,
+            pending: VecDeque::new(),
+        })
+    }
+
+    pub fn submit_bulk(&mut self, endpoint: u8, buf: Vec<u8>) -> Result<()> {
+        // Safety: If transfer is submitted, it is pushed onto `pending` where it will be
+        // dropped before `device` is freed.
+        unsafe {
+            let mut transfer = Transfer::bulk(self.device.as_raw(), endpoint, buf);
+            transfer.submit()?;
+            self.pending.push_back(transfer);
+            Ok(())
+        }
+    }
+
+    pub fn submit_control(
+        &mut self,
+        request_type: u8,
+        request: u8,
+        value: u16,
+        index: u16,
+        data: &[u8],
+    ) -> Result<()> {
+        // Safety: If transfer is submitted, it is pushed onto `pending` where it will be
+        // dropped before `device` is freed.
+        unsafe {
+            let mut transfer = Transfer::control(
+                self.device.as_raw(),
+                request_type,
+                request,
+                value,
+                index,
+                data,
+            );
+            transfer.submit()?;
+            self.pending.push_back(transfer);
+            Ok(())
+        }
+    }
+
+    pub unsafe fn submit_control_raw(&mut self, buffer: Vec<u8>) -> Result<()> {
+        // Safety: If transfer is submitted, it is pushed onto `pending` where it will be
+        // dropped before `device` is freed.
+        unsafe {
```

Comment:
I'm not really that knowledgeable in unsafe rust, but when I patched in a modified version while trying to fix ISO, rustc warned me that an `unsafe {` block is not necessary because it's an `unsafe fn`